Sources/TensorFlow/Operators/NN.swift

```diff
@@ -138,434 +137,421 @@ public extension Padding {
-        return (value, { v in
-            (self.conv2DBackpropFilter(
-                input: v, filterSizes: shape, strides: strides, padding: padding),
-             v.convolved2D(withFilter: filter, strides: strides, padding: padding))
-        })
-    }
-
-    @inlinable
-    internal func _vjpConv2DBackpropFilter(
-        _ input: Tensor,
-        _ filterSizes: Tensor<Int32>,
-        _ strides: (Int, Int, Int, Int),
-        _ padding: Padding
-    ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
-        let value = conv2DBackpropFilter(
-            input: input, filterSizes: filterSizes, strides: strides, padding: padding)
-        return (value, { v in
-            (self.conv2DBackpropInput(
-                shape: filterSizes, filter: v, strides: strides, padding: padding),
-             input.convolved2D(withFilter: v, strides: strides, padding: padding))
-        })
-    }
-
-    @inlinable
-    internal func _vjpConvolved2D(
-        filter: Tensor,
-        strides: (Int, Int, Int, Int),
-        padding: Padding
-    ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
-        let value = convolved2D(withFilter: filter, strides: strides, padding: padding)
-        return (value, { v in
-            (v.conv2DBackpropInput(
-                shape: self.shapeTensor, filter: filter, strides: strides, padding: padding),
-             v.conv2DBackpropFilter(
-                input: self, filterSizes: filter.shapeTensor, strides: strides, padding: padding))
-        })
-    }
+/// TensorFlow builtin conv2d gradient helper for the input.
+@differentiable(wrt: (x, filter), vjp: _vjpConv2DBackpropInput)
+public func conv2DBackpropInput<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>,
+    shape: Tensor<Scalar><Int32>,
+    filter: Tensor<Scalar>,
+    strides: (Int, Int, Int, Int),
+    padding: Padding
+) -> Tensor<Scalar> {
+    return Raw.conv2DBackpropInput(
+        inputSizes: shape,
+        filter: filter,
+        outBackprop: x,
+        strides: [Int32(strides.0), Int32(strides.1), Int32(strides.2), Int32(strides.3)],
+        padding: padding.raw2,
+        explicitPaddings: [])
+}
 
-    /// TensorFlow builtin conv3d gradient helper for the input.
-    @inlinable
-    @differentiable(wrt: (self, filter), vjp: _vjpConv3DBackpropInput)
-    internal func conv3DBackpropInput(
-        shape: Tensor<Int32>,
-        filter: Tensor,
-        strides: (Int, Int, Int, Int, Int),
-        padding: Padding
-    ) -> Tensor {
-        return Raw.conv3DBackpropInputV2(
-            inputSizes: shape,
-            filter: filter,
-            outBackprop: self,
-            strides: [Int32(strides.0), Int32(strides.1), Int32(strides.2),
-                      Int32(strides.3), Int32(strides.4)],
-            padding: padding.raw)
-    }
+/// TensorFlow builtin conv2d gradient helper for the filter.
+@differentiable(wrt: (x, input), vjp: _vjpConv2DBackpropFilter)
+public func conv2DBackpropFilter<Scalar: TensorFlowFloatingPoint>(
```

Comment:
This function is not a public API (it's originally `internal`). Could you please drop the `public` attribute and add `@usableFromInline`?
```suggestion
@usableFromInline
func conv2DBackpropFilter<Scalar: TensorFlowFloatingPoint>(
```